sphinx/source/docs/dev_guide/writing_tests.rst

```diff
@@ -0,0 +1,367 @@
+resembles the structure of Bokeh's Python models.
+
+Python integration tests use pytest fixtures to handle the webdriver
+configuration and interaction with Selenium. Depending on which context you
+want to test an object in, choose from ``bokeh_model_page``,
+``single_plot_page``, or ``bokeh_server_page``. See
+:bokeh-tree:`bokeh/_testing/plugins/project.py` for more details.
+
+Follow these guidelines when adding or updating Python integration tests:
+
+Keep your code as simple as possible
+    Try to only include things that are essential to your test. Focus your test
+    on one specific functionality. If possible, write several small tests
+    instead of one complex one.
+
+Use the |bokeh.models| API whenever possible
+    Try to use Bokeh's
+    :ref:`low-level bokeh.models interface <userguide_interfaces_models>` instead of
+    the more high-level
+    :ref:`bokeh.plotting interface <userguide_interfaces_plotting>`.
+
+.. _devguide_writing_tests_bokehjs:
+
+Writing JavaScript tests (BokehJS)
+----------------------------------
+
+To maintain the functionality of all :term:`BokehJS` components, Bokeh includes
+various tests written in TypeScript. If all or parts of your changes affect
+BokehJS' JavaScript code, you should add or update the relevant BokehJS tests.
+
+Bokeh's JavaScript tests use a custom testing framework that **requires Google
+Chrome or Chromium**. You need a recent version of one of these browsers
+available on your system to work with these tests.
+
+Like several other testing frameworks such as `Mocha`_ or `Jasmine`_, the
+BokehJS testing framework uses ``describe()`` and ``it()`` functions to set up
+tests.
+
+The BokehJS tests are located in :bokeh-tree:`bokehjs/test`. See
+:ref:`devguide_testing_local_javascript` for information on how to run them.
+
+Information on contributing to BokehJS is available in
+:ref:`devguide_bokehjs`.
+
+.. _devguide_writing_tests_bokehjs_unit:
+
+BokehJS unit tests
+~~~~~~~~~~~~~~~~~~
+
+The :term:`BokehJS` unit tests help make sure that the individual sections of
+BokehJS function as expected. The unit tests for BokehJS are located in the
+:bokeh-tree:`bokehjs/test/unit/` folder and sub-folders.
+
+Use `Chai "expect" assertion style <Chai_>`_ when writing unit tests for the
+BokehJS testing framework. See the `API documentation of the Chai Assertion
+Library <Chai documentation_>`_ for more details on this style.
+
+Use ``expect()`` together with the following elements to create assertions for
+the BokehJS testing framework:
+
+* ``to`` and ``be``: tokens to improve readability of assertions and connect
+  elements
+* ``not``: negates the following assertions
+* ``throw``: asserts that an error is thrown. Accepts the following optional
+  parameters: ``error_type`` (filter by ``Error``) and ``pattern`` (filter by
+  regular expression or string).
+* ``equal``: asserts strict equality (``===``). Expects an operand to compare
```

Comment:
Equality is defined as (deep) value equality, instead of reference equality given by `==` or `===`. For reference equality one would use `identical` method (strict, i.e. `===`). Value equality is provided by `core/util/eq` module, thus is the same across bokehjs (e.g. properties use it to detect change).